File: backend/app/services/hope_adapter.py
"""
Adapter Hope/Titan → formato interno Proposta.

Hope é um banco específico integrado via API Titan (Ceoslab).
Este é o ÚNICO lugar onde banco = "HOPE" é hardcodado — correto por design,
pois Titan/Ceoslab é exclusivamente o sistema do banco Hope.

Consolida a lógica de _map_operacao() de titan_sync.py em camada reutilizável.
"""
from __future__ import annotations

import re
from typing import Any


# ── Helpers ───────────────────────────────────────────────────────────────────

def _str(*candidates: Any) -> str:
    for v in candidates:
        if v is None or v == "":
            continue
        if isinstance(v, str):
            s = v.strip()
            if s:
                return s
        if isinstance(v, (int, float)):
            return str(v)
    return ""


def _cpf_digits(v: str | None) -> str | None:
    if not v:
        return None
    digits = re.sub(r"\D", "", v)
    return digits if len(digits) in (11, 14) else None


def _first_address(addresses: list) -> dict:
    if addresses and isinstance(addresses[0], dict):
        return addresses[0]
    return {}


# ── Mapeamento principal ──────────────────────────────────────────────────────

def mapear_operacao(op: dict) -> dict | None:
    """
    Converte uma operação da API Titan para o formato PropostaCreate interno.

    banco = "HOPE" é INTENCIONAL aqui — Titan é o sistema exclusivo do
    banco Hope. Este é o único local do código onde esse hardcode é válido.

    Retorna None se CPF, valor ou ID estiverem ausentes.
    """
    op_id = op.get("id")
    if not op_id:
        return None

    customer    = op.get("customer") or {}
    person      = customer.get("person") or {}
    product     = op.get("product") or {}
    originating = op.get("originatingCompany") or {}
    addresses   = person.get("addresses") or []
    address     = _first_address(addresses)

    cpf = _cpf_digits(person.get("documentNumber"))
    if not cpf:
        return None

    valor = op.get("requestedValue") or 0.0
    if float(valor) <= 0:
        return None

    return {
        "proposta_id_externo": f"titan-{op_id}",
        "cpf_cliente": cpf,
        "nome_cliente": _str(person.get("fullName")) or None,
        "uf_cliente": _str(address.get("state")) or None,
        "banco": "HOPE",            # INTENCIONAL — Titan/Ceoslab = banco Hope
        "convenio": _str(originating.get("tradeName")) or None,
        "produto": _str(product.get("name")) or None,
        "valor": float(valor),
        "payload_original": op,
    }

File: frontend/src/lib/hopeAdapter.ts
/**
 * hopeAdapter — frontend adapter para a API Hope/Titan (Ceoslab).
 *
 * Hope é um banco específico, integrado exclusivamente via API Titan.
 * banco = "HOPE" é correto e intencional aqui — único local válido
 * no frontend (assim como hope_adapter.py no backend).
 */

import api from "@/lib/api";

// ── Tipos da API Titan ────────────────────────────────────────────────────────

export interface TitanAddress {
  street?: string;
  number?: string;
  city?: string;
  state?: string;
  zipCode?: string;
}

export interface TitanPerson {
  fullName?: string;
  documentNumber?: string;
  birthDate?: string;
  gender?: string;
  addresses?: TitanAddress[];
}

export interface TitanCustomer {
  person?: TitanPerson;
}

export interface TitanProduct {
  id?: number;
  name?: string;
  type?: string;
}

export interface TitanOriginatingCompany {
  id?: number;
  tradeName?: string;
  cnpj?: string;
}

export interface TitanOperacao {
  id: string | number;
  requestedValue?: number;
  status?: string;
  createdAt?: string;
  updatedAt?: string;
  customer?: TitanCustomer;
  product?: TitanProduct;
  originatingCompany?: TitanOriginatingCompany;
}

// ── Normalização para exibição ────────────────────────────────────────────────

export interface OperacaoNorm {
  id: string;
  ade: string;
  banco: "HOPE";           // INTENCIONAL — Titan = banco Hope exclusivamente
  convenio: string;
  produto: string;
  cpf: string;
  nome: string;
  uf: string;
  valor: string;
  valor_raw: number | null;
  criado_em: string;
}

function _str(v: unknown): string {
  if (!v) return "";
  return typeof v === "string" ? v.trim() : String(v);
}

function _cpfDigits(v: string): string {
  const d = v.replace(/\D/g, "");
  if (d.length === 11) return `${d.slice(0,3)}.${d.slice(3,6)}.${d.slice(6,9)}-${d.slice(9)}`;
  return d;
}

function _fmtBRL(v: number): string {
  return v.toLocaleString("pt-BR", { style: "currency", currency: "BRL" });
}

/**
 * Normaliza uma operação Titan para exibição no frontend.
 * banco sempre "HOPE" — Titan é o sistema exclusivo do banco Hope.
 */
export function normOperacao(op: TitanOperacao): OperacaoNorm {
  const person      = op.customer?.person ?? {};
  const originating = op.originatingCompany ?? {};
  const product     = op.product ?? {};
  const addresses   = person.addresses ?? [];
  const address     = addresses[0] ?? {};

  const valor_raw = op.requestedValue ? Number(op.requestedValue) : null;

  return {
    id:        _str(op.id),
    ade:       `titan-${_str(op.id)}`,
    banco:     "HOPE",
    convenio:  _str(originating.tradeName),
    produto:   _str(product.name),
    cpf:       _cpfDigits(_str(person.documentNumber)),
    nome:      _str(person.fullName),
    uf:        _str(address.state),
    valor:     valor_raw != null ? _fmtBRL(valor_raw) : "—",
    valor_raw,
    criado_em: _str(op.createdAt),
  };
}

// ── Sincronização ─────────────────────────────────────────────────────────────

export interface TitanSyncParams {
  page_size?: number;
  max_pages?: number;
  data_inicio?: string;
  data_fim?: string;
  status_id?: number;
}

export interface TitanSyncResult {
  importadas: number;
  ignoradas: number;
  erros: number;
}

/**
 * Dispara a sincronização Hope/Titan → propostas no backend.
 * Chama POST /titan/sync e retorna os contadores.
 */
export async function titanSync(params: TitanSyncParams = {}): Promise<TitanSyncResult> {
  const query = new URLSearchParams();
  if (params.page_size   != null) query.set("page_size",   String(params.page_size));
  if (params.max_pages   != null) query.set("max_pages",   String(params.max_pages));
  if (params.data_inicio)         query.set("data_inicio", params.data_inicio);
  if (params.data_fim)            query.set("data_fim",    params.data_fim);
  if (params.status_id   != null) query.set("status_id",   String(params.status_id));

  const qs = query.toString();
  const url = `/titan/sync${qs ? `?${qs}` : ""}`;
  const { data } = await api.post<TitanSyncResult>(url);
  return data;
}
